predicators/approaches/grammar_search_invention_approach.py

```diff
@@ -442,6 +498,88 @@ def _get_feature_ranges(
         return feature_ranges
 
 
+@dataclass(frozen=True, eq=False, repr=False)
+class _DoubleFeatureInequalitiesPredicateGrammar(
+        _SingleFeatureInequalitiesPredicateGrammar):
+    """Generates features of the form "|0.feature - 1.feature| >= c" or
+    "|0.feature - 1.feature| <= c". This is also capable of
```

Comment:
I'm finding the fact that this grammar also produces single feature inequalities a bit confusing, and redundant with the code we have for single feature inequalities -- it might be the best option, but let's talk through an alternative.

could we instead have this grammar only produce double (difference) feature predicates, and then "chain" together this grammar and the original single features grammar? we already have a `_ChainPredicateGrammar`, but as implemented right now, it won't work, because it enumerates all predicates from the first grammar before moving onto the second one, and both of these grammars are infinite. so instead, we could have the chaining alternate between the grammars, like this: https://stackoverflow.com/questions/2017895/alternating-between-iterators-in-python we could make "alternate" a kwarg for `_ChainPredicateGrammar`. would this work?

Reply:
in the stackoverflow answers, I think the "roundrobin" one is probably the best

Reply:
i agree that it's confusing this grammar produces single feature inequalities; let me look into this alternative and see if I can get it to work!

Reply:
got it working!